### PR TITLE
Add font-hack-nerd-font [WIP] [RFC]

### PR DESCRIPTION
**Pull request by ryanoasis**
A WIP & RFC only PR to begin an actual later PR to include all of the patched fonts from https://github.com/ryanoasis/nerd-fonts

After making all changes to the cask:

- [ ] `brew cask audit --download {{cask_file}}` is error-free. __*no, see below__
- [ ] `brew cask style --fix {{cask_file}}` reports no offenses. __*no, see below__
- [x] The commit message includes the cask’s name and version.

Additionally, **if adding a new cask**:

- [x] Named the cask according to the [token reference].
- [ ] `brew cask install {{cask_file}}` worked successfully. __*no, see below__
- [ ] `brew cask uninstall {{cask_file}}` worked successfully. __*no, see below__
- [x] Checked there are no [open pull requests] for the same cask.
- [x] Checked that the cask was not already refused in [closed issues].

[token reference]: https://github.com/caskroom/homebrew-cask/blob/master/doc/cask_language_reference/token_reference.md
[open pull requests]: https://github.com/caskroom/homebrew-fonts/pulls
[closed issues]: https://github.com/caskroom/homebrew-fonts/issues?q=is%3Aissue+is%3Aclosed

* *__did not check all the requirements because I cannot test on macOS, apologies:__
  * What I have looks like it _should_ work and is based on discussions on https://github.com/ryanoasis/nerd-fonts/issues/72 including @vitorgalvao
  * Of course feel free to deny this PR if you want me or someone else to test it but was hoping to get help regarding that here :)
* Just adding Hack for RFC WIP PR (eventually want to add ~32 casks)
* source and discussion reference for the PR: https://github.com/ryanoasis/nerd-fonts/issues/72
* cask generated with script from [here](https://github.com/ryanoasis/nerd-fonts/blob/0.9.0/bin/scripts/generate-casks.sh)

### CODE REVIEW

File: Casks/font-hack-nerd-font.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+cask 'font-hack-nerd-font' do
+  version :latest
+  sha256 :no_check
+  url 'https://raw.githubusercontent.com/ryanoasis/nerd-fonts/master/patched-fonts/Hack'
```
Comment (vitorgalvao) [inline Casks/font-hack-nerd-font.rb]: Not sure what you’re pointing to, but this does not exist.

File: Casks/font-hack-nerd-font.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+  font 'Regular/complete/Knack Regular Nerd Font Complete.ttf'
+  font 'Regular/complete/Knack Regular Nerd Font Complete Mono.ttf'
+  font 'Italic/complete/Knack Italic Nerd Font Complete Mono.ttf'
+  font 'Italic/complete/Knack Italic Nerd Font Complete.ttf'
+  font 'BoldItalic/complete/Knack Bold Italic Nerd Font Complete Mono.ttf'
+  font 'BoldItalic/complete/Knack Bold Italic Nerd Font Complete.ttf'
+  font 'Bold/complete/Knack Bold Nerd Font Complete.ttf'
+  font 'Bold/complete/Knack Bold Nerd Font Complete Mono.ttf'
+
```
Comment (vitorgalvao) [inline Casks/font-hack-nerd-font.rb]: This empty line must be removed.

File: Casks/font-hack-nerd-font.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
+cask 'font-hack-nerd-font' do
+  version :latest
+  sha256 :no_check
+  url 'https://raw.githubusercontent.com/ryanoasis/nerd-fonts/master/patched-fonts/Hack'
+  homepage 'https://github.com/ryanoasis/nerd-fonts'
```
Comment (vitorgalvao) [inline Casks/font-hack-nerd-font.rb]: `homepage` need to be placed *under* `name`.

File: Casks/font-hack-nerd-font.rb (hunk 1 above)
Comment (vitorgalvao) [inline Casks/font-hack-nerd-font.rb]: There needs to be an empty line above `url`.

Comment (ryanoasis) [inline Casks/font-hack-nerd-font.rb]: So I wasn't sure exactly how the process works what url is supposed to point to. I thought it gets used in conjuction with the `font` stanza lines to build a URL for download or something.

Basically was trying to point to this: https://github.com/ryanoasis/nerd-fonts/tree/master/patched-fonts/Hack

Does that mean if I have fonts in separate folders I have to set the `font` lines to the entire path? Such as:

```rb
  font 'https://github.com/ryanoasis/nerd-fonts/blob/master/patched-fonts/Hack/Regular/complete/Knack%20Regular%20Nerd%20Font%20Complete.ttf?raw=true'
  font 'https://github.com/ryanoasis/nerd-fonts/blob/master/patched-fonts/Hack/Regular/complete/Knack%20Regular%20Nerd%20Font%20Complete%20Mono.ttf?raw=true'
  #  ...
```

File: Casks/font-hack-nerd-font.rb (hunk 3 above)
Comment (ryanoasis) [inline Casks/font-hack-nerd-font.rb]: Thanks. I will fix it.

File: Casks/font-hack-nerd-font.rb (hunk 1 above)
Comment (vitorgalvao) [inline Casks/font-hack-nerd-font.rb]: No, we don’t do any hidden magic. What you see is what we use.

`url` is a literal URL for something to download (usually a package/archive, like a `dmg` or `zip`). The `font` stanzas indicate the paths relative to that archive where the files are, they cannot point to URLs (the `font` stanzas you have now make sense).

In this case, since you don’t have archived versions of the fonts, you’d basically need to have one cask for each font file, which would be ridiculous (and we wouldn’t accept). You could also like to github’s master archive for the repo, but in this case that would include so many unrelated fonts we wouldn’t accept it as well.

Basically the way to go here is to make an archive for each font family.

Comment (ryanoasis) [inline Casks/font-hack-nerd-font.rb]: Okay actually that is one of my initial doubts is that this would require an archive file. Sorry for the confusion.

Just to be clear are you saying if I used the same github master archive for each font and limited the `font` stanzas to just the relevant ones that would not be accepted?

I will go forward with creating individual font archives as you recommend in the meantime.

Thanks!

Comment (vitorgalvao) [inline Casks/font-hack-nerd-font.rb]: > Just to be clear are you saying if I used the same github master archive for each `font` and limited the font stanzas to just the relevant ones that would not be accepted?

Yes, that is correct. The master archive for that repo is over 1GB (not even sure how over, I stopped around there) which is ***way*** too much if you only want one font family and way larger than the vast majority of apps, even. It would be akin (although on a smaller scale) to downloading the entire Google Fonts repo to get a single font, which would be insane.

Comment (ryanoasis) [inline Casks/font-hack-nerd-font.rb]: Yes I understand and agree with you. Actually the repo size is going to be addressed but still would be quite large.. just not ridiculously so :stuck_out_tongue: 

I am going to move forward with archives per font (For reference: https://github.com/ryanoasis/nerd-fonts/issues/32)

File: Casks/font-hack-nerd-font.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
+cask 'font-hack-nerd-font' do
+  version 'v0.8.0'
+  sha256 '89aff63d5acbe582e11895e745b6da78a75fa2007552c7a1be1c9cc35296eb59'
+
+  url 'https://github.com/ryanoasis/nerd-fonts/releases/download/v0.8.0/Hack.zip'
```
Comment (vitorgalvao) [inline Casks/font-hack-nerd-font.rb]: Please add an [`appcast`](https://github.com/caskroom/homebrew-cask/blob/master/doc/cask_language_reference/stanzas/appcast.md).

Comment (vitorgalvao) [inline Casks/font-hack-nerd-font.rb]: `url "https://github.com/ryanoasis/nerd-fonts/releases/download/v#{version}/Hack.zip"`

File: Casks/font-hack-nerd-font.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
+cask 'font-hack-nerd-font' do
+  version 'v0.8.0'
```
Comment (vitorgalvao) [inline Casks/font-hack-nerd-font.rb]: `version '0.8.0'`

**Comment by vitorgalvao** [comment]
You added some irrelevant commits to the PR.

**Comment by ryanoasis** [comment]
@vitorgalvao Yeah I noticed the commits, not sure what I did, I am going to try to straighten things out today. Thanks

File: Casks/font-hack-nerd-font.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+  font 'Knack Regular Nerd Font Complete.ttf'
+  font 'Knack Regular Nerd Font Complete Mono.ttf'
+  font 'Knack Italic Nerd Font Complete Mono.ttf'
+  font 'Knack Italic Nerd Font Complete.ttf'
+  font 'Knack Bold Italic Nerd Font Complete Mono.ttf'
+  font 'Knack Bold Italic Nerd Font Complete.ttf'
+  font 'Knack Bold Nerd Font Complete.ttf'
+  font 'Knack Bold Nerd Font Complete Mono.ttf'
+end
```
Comment (vitorgalvao) [inline Casks/font-hack-nerd-font.rb]: It has a “no newline at end of file” warning. Those are [actually meaningful](http://unix.stackexchange.com/a/18789). Apart from that, it looks fine.

File: Casks/font-hack-nerd-font.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
+cask 'font-hack-nerd-font' do
+  version '0.8.0'
+  sha256 '89aff63d5acbe582e11895e745b6da78a75fa2007552c7a1be1c9cc35296eb59'
+
+  url 'https://github.com/ryanoasis/nerd-fonts/releases/download/v#{version}/Hack.zip'
+  appcast 'https://github.com/ryanoasis/nerd-fonts/releases.atom'
```
Comment (vitorgalvao) [inline Casks/font-hack-nerd-font.rb]: You forgot a comma at the end of this line. That’s why the build is failing.

File: Casks/font-hack-nerd-font.rb [hunk 8]
```diff
@@ -0,0 +1,19 @@
+cask 'font-hack-nerd-font' do
+  version '0.8.0'
+  sha256 '89aff63d5acbe582e11895e745b6da78a75fa2007552c7a1be1c9cc35296eb59'
+
+  url 'https://github.com/ryanoasis/nerd-fonts/releases/download/v#{version}/Hack.zip'
+  appcast 'https://github.com/ryanoasis/nerd-fonts/releases.atom',
```
Comment (vitorgalvao) [inline Casks/font-hack-nerd-font.rb]: You forgot a comma at the end of this line. That’s why the build is failing.

I’ve added the line this time, but that needs to be fixed in your generator.

Comment (ryanoasis) [inline Casks/font-hack-nerd-font.rb]: Whoops sorry. Thanks for the fix. I have updated my script :smile:

File: Casks/font-hack-nerd-font.rb [hunk 9]
```diff
@@ -0,0 +1,19 @@
+cask 'font-hack-nerd-font' do
+  version '0.8.0'
+  sha256 '89aff63d5acbe582e11895e745b6da78a75fa2007552c7a1be1c9cc35296eb59'
+
+  url "https://github.com/ryanoasis/nerd-fonts/releases/download/v#{version}/Hack.zip"
```
Comment (vitorgalvao) [inline Casks/font-hack-nerd-font.rb]: When you use `#{version}` (i.e. interpolation), you need double quotes. Fixed it now, but also needs to be fixed in your script.

**Comment by vitorgalvao** [comment]
I’d suggest you submit a new one using your revised script. If it’s fine, then submit all the others.

**Comment by ryanoasis** [comment]
@vitorgalvao Yes sure I will submit one more PR before submitting a PR with the remaining fonts
